### PR TITLE
Change GridDeviceMetadata qubit type to GridQubit

### PR DESCRIPTION
**Pull request by verult**
Proposal to change qubit types in `GridDeviceMetadata` to `GridQubit`, per @maffoo 's and @mpharrigan 's suggestion.

Looks like `GridQid` is not a superclass of `GridQubit`.

Also included a quick docstring fix to disambiguate from `cirq_google.GridDevice`.

@MichaelBroughton WDYT?

### CODE REVIEW

**Comment by MichaelBroughton** [comment]
Seems like a good idea to me!

**Review by mpharrigan — APPROVED** [review]
do you want to remove the casts that someone linked to in our chat in a recent @MichaelBroughton pr?

**Comment by verult** [comment]
@mpharrigan Good point, will fix all the casts used in the codebase.

**Comment by verult** [comment]
`SerializableDevice` technically supports both GridQubits and NamedQubits. Since it's being deprecated and all Cirq call sites of `SerializableDevice` has already been rerouted, I did the simplest thing of breaking qubit types other than GridQubits. Let me know if there are any breakages in downstream dependencies and I'm happy to help fix them.

**Review by pavoljuhas — APPROVED** [review]

File: cirq-google/cirq_google/devices/serializable_device.py [hunk 1]
```diff
@@ -101,15 +101,22 @@ def __init__(
...
+            qubit_pairs=cast(
+                List[Tuple[cirq.GridQubit, cirq.GridQubit]],
+                [
+                    (pair[0], pair[1])
+                    for gate_defs in gate_definitions.values()
+                    for gate_def in gate_defs
+                    if gate_def.number_of_qubits == 2
+                    for pair in gate_def.target_set
+                    if len(pair) == 2
+                    and pair[0] < pair[1]
```
Comment (mpharrigan) [inline cirq-google/cirq_google/devices/serializable_device.py:115]: I know this was already here, but does this mean if you order your qubits "backwards" in target_set the metadata will report that there are no pairs?

File: cirq-google/cirq_google/devices/serializable_device.py [hunk 2]
```diff
@@ -101,15 +101,22 @@ def __init__(
...
+                    and isinstance(pair[0], cirq.GridQubit)
+                    and isinstance(pair[1], cirq.GridQubit)
```
Comment (mpharrigan) [inline cirq-google/cirq_google/devices/serializable_device.py:117]: doesn't this obviate the need for the cast? Usually mypy can use `if isinstance(...)` checks to infer variable types

File: cirq-google/cirq_google/devices/serializable_device.py [hunk 3]
```diff
@@ -101,15 +101,22 @@ def __init__(
...
-                for gate_defs in gate_definitions.values()
-                for gate_def in gate_defs
-                if gate_def.number_of_qubits == 2
-                for pair in gate_def.target_set
-                if len(pair) == 2 and pair[0] < pair[1]
-            ],
+            qubit_pairs=cast(
+                List[Tuple[cirq.GridQubit, cirq.GridQubit]],
+                [
+                    (pair[0], pair[1])
```
Comment (mpharrigan) [inline cirq-google/cirq_google/devices/serializable_device.py:109]: put the cast here so you're casting less things? Like it still can infer the `List[Tuple[...` part; you need to cast the qubits themselves. Although with the `isinstance` checks I'm sad that that's necessary, see below.

File: cirq-google/cirq_google/line/placement/greedy.py [hunk 4]
```diff
@@ -304,7 +304,7 @@ def place_line(self, device: 'cirq_google.GridDevice', length: int) -> GridQubit
         if not device.metadata.qubit_set:
             return GridQubitLineTuple()
 
-        start: GridQubit = cast(GridQubit, min(device.metadata.qubit_set))
```
Comment (mpharrigan) [inline cirq-google/cirq_google/line/placement/greedy.py:307]: nice

File: cirq-core/cirq/devices/grid_device_metadata.py [hunk 5]
```diff
@@ -114,7 +114,16 @@ def __init__(
...
 
     @property
-    def qubit_pairs(self) -> FrozenSet[FrozenSet['cirq.Qid']]:
+    def qubit_set(self) -> FrozenSet['cirq.GridQubit']:
+        """Returns the set of grid qubits on the device.
+
+        Returns:
+            Frozenset of qubits on device.
+        """
+        return cast(FrozenSet['cirq.GridQubit'], super().qubit_set)
```
[the remark this answers is not in the document]
Comment (verult) [reply cirq-core/cirq/devices/grid_device_metadata.py:123]: Right. `self.qubits` doesn't exist, and I wasn't sure about adding a `self._qubits` attribute just for this when it's already accessible from the superclass but just with a more general type.

File: cirq-google/cirq_google/devices/serializable_device.py (hunk 2 above)
Comment (verult) [reply cirq-google/cirq_google/devices/serializable_device.py:117]: I was hoping for that too, but unfortunately mypy complained (that the type is `List[Tuple[Qid, Qid]]`)

File: cirq-google/cirq_google/devices/serializable_device.py (hunk 1 above)
Comment (verult) [reply cirq-google/cirq_google/devices/serializable_device.py:115]: Ah good catch. This was added when `GridDeviceMetadata` required the pair to be ordered but missed this call site when it was changed to be order-agnostic. Will fix separately.

**Review by maffoo — APPROVED** [review]

**Pull request merged**